enhancements/ingress/set-replace-append-delete-deny-headers.md

```diff
@@ -0,0 +1,993 @@
+    Defining validation for the same API in both library-go and the OpenAPI schema introduces the risk that someone could inadvertently introduce discrepancies between the library-go validation and the OpenAPI schema, leading to buggy and potentially confusing behavior.  Care must be taken to avoid introducing such discrepancies.
+    Historical note: At one time, MicroShift relied on the OpenAPI schema for server-side validation of the Route CRD, and for this reason, kubebuilder validation tags were added to the
+    Route CRD (see <https://github.com/openshift/api/pull/1228>).  Since then, an admission plugin has been added to kube-apiserver that validates the Route CRD using the same
+    library-go validation on MicroShift that openshift-apiserver uses on OpenShift (see <https://github.com/openshift/kubernetes/pull/1416>).  As a result, the kubebuilder validation tags
+    on the Route CRD are no longer required specifically for MicroShift; however, the kubebuilder validation tags are still required for client-side validation.
+
+- Why don't you set/delete the headers on frontend public_ssl in the haproxy.config?
+    Answer:
+    The connection is still encrypted in public_ssl, so it is not possible to examine or modify the HTTP session there. The fe_sni and fe_no_sni frontends terminate TLS.
+
+
+### Test Plan
+
+The controller that manages the IngressController Deployment and related
+resources has unit test coverage; for this enhancement, the unit tests are
+expanded to cover the additional functionality.
+Unit tests for testing regular expression to test invalid values will be written in the library-go.
+Unit test to check the deployment env var having base64 encoded value of the router deployment will be written in the ingress operator.
+Unit tests to check if the values are passed correctly to the template will be written in the router.
+Unit tests to test escaping and quoting will be written in the library-go.
+
+The operator has end-to-end tests; for this enhancement, a test is added that
+(1) creates an IngressController that specifies certain response
+headers should be set or deleted, (2) sends an HTTP request to one of the standard
+routes (such as the console route), and (3) verifies that the expected headers
+are seen in the `curl -vk <route host>`.
+
+Integration tests will be added in openshift/api.
+
+### Graduation Criteria
+
+#### Dev Preview -> Tech Preview
+
+#### Tech Preview -> GA
+
+#### Removing a deprecated feature
+
+### Upgrade / Downgrade Strategy
+
+On upgrade, the default configuration sets or deletes no HTTP headers, as before the
+introduction of the HTTP header Set/Delete API.
+
+On downgrade, the HAProxy configuration template ignores unrecognized
+environment variables set by ingress operator using the value set in the Ingress Controller CR and
+the ServiceAlias config struct variable set by router using the value set in the routeSpec.
```

Comment:
How can the `ServiceAliasConfig` struct provide unrecognized values to the template?  The `ServiceAliasConfig` struct and `haproxy-config.template` are defined in the same container, so the versions should always match.

Reply:
i added a different text.